primitives/xcm/src/lib.rs

```diff
@@ -250,3 +251,56 @@ impl<
         }
     }
 }
+
+/// Allows execution from `origin` if it is contained in `T` (i.e. `T::Contains(origin)`) taking
+/// payments into account.
+///
+/// Only allows for sequence `DescendOrigin` -> `WithdrawAsset` -> `BuyExecution`
+pub struct AllowPaidExecWithDescendOriginFrom<T>(PhantomData<T>);
+impl<T: Contains<MultiLocation>> ShouldExecute for AllowPaidExecWithDescendOriginFrom<T> {
+    fn should_execute<RuntimeCall>(
+        origin: &MultiLocation,
+        message: &mut Xcm<RuntimeCall>,
+        max_weight: Weight,
+        _weight_credit: &mut Weight,
+    ) -> Result<(), ()> {
+        log::trace!(
+            target: "xcm::barriers",
+            "AllowPaidExecWithDescendOriginFrom origin: {:?}, message: {:?}, max_weight: {:?}, weight_credit: {:?}",
+            origin, message, max_weight, _weight_credit,
+        );
+        ensure!(T::contains(origin), ());
+        let mut iter = message.0.iter_mut();
+        let mut i = iter.next().ok_or(())?;
+
+        match i {
```

Comment:
If I'm not mistaken, this whole sequence can be replaced with a single `assert_matches`. Or, if it's not available in this code, a match on an array.

Something like this:
```rust
match iter.collect::<Vec<_>>() {
    [DescendOrigin(..), WithdrawAsset(..), BuyExecution { weight_limit, .. } ] => { /* check weight_limit */ }
    _ => return Err(())
}
```

Reply:
I haven't seen `assert_matches!` used outside of unit test context so would like to avoid it in runtime code.

For the suggestion you made, I did it like this:
```Rust
        let iter = message.0.iter_mut();

        match iter.take(3).collect::<Vec<_>>()[..] {
            [DescendOrigin(..), WithdrawAsset(..), BuyExecution {
                ref mut weight_limit,
                ..
            }] => {
                match weight_limit {
                    Limited(ref mut weight) if *weight >= max_weight => {
                        *weight = max_weight;
                        Ok(())
                    }
                    Unlimited => {
                        *weight_limit = Limited(max_weight);
                        Ok(())
                    }
                    _ => Err(()),
                }
            }
            _ => return Err(()),
        }
```
I feel as if few extra lines were saved but readability has decreased significantly.

Reply:
Agree. This version do look awkward. I missed that we need to assign the `weight_limit` back.

Reply:
Out of curiosily, would something like this work? Don't sure if it would compile, though.

```rust
        match iter.take(3).collect::<Vec<_>>()[..] {
            [DescendOrigin(..), WithdrawAsset(..), BuyExecution { weight_limit: Limited(ref mut limit), .. }] if *limit >= max_weight => {
                *limit = max_weight;
                Ok(())
            }

            [DescendOrigin(..), WithdrawAsset(..), BuyExecution { weight_limit: Unlimited(ref mut limit), .. }] => {
                *weight_limit = Limited(max_weight);
                Ok(())
            }

            _ => return Err(()),
        }    
```

Reply:
Comments implemented.

Also, paranoia kicked in so I added additional UTs.

Reply:
@0x7CFE if no more comments, can you please approve?